File: clients/javascript/src/element-list.ts
import { BiDiClient } from './bidi';
import { Element, ElementInfo, SelectorOptions } from './element';

export interface ElementListInfo {
  tag: string;
  text: string;
  box: { x: number; y: number; width: number; height: number };
  index: number;
}

export interface FilterOptions {
  hasText?: string;
  has?: string;
}

export class ElementList {
  private client: BiDiClient;
  private context: string;
  private selector: string | SelectorOptions;
  private elements: Element[];

  constructor(
    client: BiDiClient,
    context: string,
    selector: string | SelectorOptions,
    elements: Element[]
  ) {
    this.client = client;
    this.context = context;
    this.selector = selector;
    this.elements = elements;
  }

  /** Number of elements in the list. */
  count(): number {
    return this.elements.length;
  }

  /** First element in the list. */
  first(): Element {
    return this.nth(0);
  }

  /** Last element in the list. */
  last(): Element {
    return this.nth(this.elements.length - 1);
  }

  /** Element at the given index. */
  nth(index: number): Element {
    if (index < 0 || index >= this.elements.length) {
      throw new Error(`Index ${index} out of bounds (0..${this.elements.length - 1})`);
    }
    return this.elements[index];
  }

  /** Filter elements by re-querying with filter params. */
  async filter(opts: FilterOptions): Promise<ElementList> {
    const params: Record<string, unknown> = {
      context: this.context,
      timeout: 5000,
    };

    // Merge original selector params
    if (typeof this.selector === 'string') {
      params.selector = this.selector;
    } else {
      Object.assign(params, this.selector);
    }

    if (opts.hasText) params.hasText = opts.hasText;
    if (opts.has) params.has = opts.has;

    const result = await this.client.send<{
      elements: ElementListInfo[];
      count: number;
    }>('vibium:findAll', params);

    const elements = result.elements.map((el) => {
      const info: ElementInfo = { tag: el.tag, text: el.text, box: el.box };
      const selectorStr = typeof this.selector === 'string' ? this.selector : '';
      const selectorParams = typeof this.selector === 'string' ? { selector: this.selector } : { ...this.selector };
      return new Element(this.client, this.context, selectorStr, info, el.index, selectorParams);
    });

    return new ElementList(this.client, this.context, this.selector, elements);
  }

  /** Return elements as an array. */
  toArray(): Element[] {
    return [...this.elements];
  }

  [Symbol.iterator](): Iterator<Element> {
    let index = 0;
    const elements = this.elements;
    return {
      next(): IteratorResult<Element> {
        if (index < elements.length) {
          return { value: elements[index++], done: false };
        }
        return { value: undefined as unknown as Element, done: true };
      },
    };
  }
}

File: clients/python/src/vibium/async_api/element_list.py
"""Async ElementList class."""

from __future__ import annotations

from typing import Any, Dict, Iterator, List, Optional, Union, TYPE_CHECKING

from .._types import BoundingBox, ElementInfo

if TYPE_CHECKING:
    from ..client import BiDiClient
    from .element import Element


class ElementList:
    """A collection of elements returned by findAll."""

    def __init__(
        self,
        client: BiDiClient,
        context: str,
        selector: Any,
        elements: List[Element],
    ) -> None:
        self._client = client
        self._context = context
        self._selector = selector
        self._elements = elements

    def count(self) -> int:
        return len(self._elements)

    def first(self) -> Element:
        return self.nth(0)

    def last(self) -> Element:
        return self.nth(len(self._elements) - 1)

    def nth(self, index: int) -> Element:
        if index < 0 or index >= len(self._elements):
            raise IndexError(f"Index {index} out of bounds (0..{len(self._elements) - 1})")
        return self._elements[index]

    async def filter(
        self,
        has_text: Optional[str] = None,
        has: Optional[str] = None,
    ) -> ElementList:
        """Filter elements by re-querying with filter params."""
        from .element import Element

        params: Dict[str, Any] = {
            "context": self._context,
            "timeout": 5000,
        }
        if isinstance(self._selector, str):
            params["selector"] = self._selector
        elif isinstance(self._selector, dict):
            params.update(self._selector)

        if has_text:
            params["hasText"] = has_text
        if has:
            params["has"] = has

        result = await self._client.send("vibium:findAll", params)

        sel_str = self._selector if isinstance(self._selector, str) else ""
        sel_params = ({"selector": self._selector} if isinstance(self._selector, str)
                      else dict(self._selector) if isinstance(self._selector, dict)
                      else {})
        elements = []
        for el in result["elements"]:
            info = ElementInfo(tag=el["tag"], text=el["text"], box=BoundingBox(**el["box"]))
            elements.append(Element(self._client, self._context, sel_str, info, el.get("index"), sel_params))
        return ElementList(self._client, self._context, self._selector, elements)

    def to_array(self) -> List[Element]:
        return list(self._elements)

    def __iter__(self) -> Iterator[Element]:
        return iter(self._elements)

    def __len__(self) -> int:
        return len(self._elements)
